src/storage-operators/src/s3_oneshot_sink/parquet.rs

```diff
@@ -7,28 +7,337 @@
+    async fn start_new_file(&mut self) -> Result<(), anyhow::Error> {
+        if let Some(active_file) = self.active_file.take() {
+            active_file
+                .finish()
+                .run_in_task(|| "ParquetFile::finish")
+                .await?;
+        }
+        let object_key = self
+            .key_manager
+            .data_key(self.batch, self.next_file_index, "parquet");
+        self.next_file_index += 1;
+
+        let schema = self.builder.schema();
+        let bucket = self.key_manager.bucket.clone();
+        let sdk_config = Arc::clone(&self.sdk_config);
+        let max_file_size = self.max_file_size;
+        let new_file = ParquetFile::new(schema, bucket, object_key, sdk_config, max_file_size)
+            .run_in_task(|| "ParquetFile::new")
+            .await?;
+
+        self.active_file = Some(new_file);
+        Ok(())
+    }
+
+    /// Flush the current arrow builder to the active file. Starts a new file if the
+    /// no file is currently active or if writing the current builder record batch to the
+    /// active file would exceed the file size limit.
+    async fn flush_builder(&mut self) -> Result<(), anyhow::Error> {
+        let builder = std::mem::replace(&mut self.builder, ArrowBuilder::new(&self.desc, 1024)?);
+        let arrow_batch = builder.to_record_batch()?;
+
+        if arrow_batch.num_rows() == 0 {
+            return Ok(());
+        }
+
+        if self.active_file.is_none() {
+            self.start_new_file().await?;
+        }
+
+        if let Some(mut active_file) = self.active_file.take() {
+            let (res, arrow_batch, mut active_file) = async move {
+                let res = active_file.write_arrow_batch(&arrow_batch).await;
+                (res, arrow_batch, active_file)
+            }
+            .run_in_task(|| "ParquetFile::write_arrow_batch")
+            .await;
+            match res {
+                Ok(()) => {}
+                Err(ParquetFileError::RecordBatchExceedsMaxFileLimit) => {
+                    // The current batch would exceed the file size limit so start a new file
+                    // and write the batch to it.
+                    self.start_new_file().await?;
+                    active_file = self.active_file.take().unwrap();
+                    active_file = async move {
+                        active_file.write_arrow_batch(&arrow_batch).await?;
+                        Ok::<_, anyhow::Error>(active_file)
+                    }
+                    .run_in_task(|| "ParquetFile::write_arrow_batch")
+                    .await?;
+                }
+                Err(e) => anyhow::bail!(e),
+            }
+            self.active_file = Some(active_file);
+        }
+        Ok(())
+    }
+}
+
+/// Helper to tie the lifecycle of the `ArrowWriter` and `S3MultiPartUploader` together
+/// for a single parquet file.
+struct ParquetFile {
+    writer: ArrowWriter<Vec<u8>>,
+    uploader: S3MultiPartUploader,
```

Comment:
That's definitely an interesting idea.. yes at the moment it would be difficult since `S3MultiPartUploader` is async, but the parquet crate does provide an async version `AsyncArrowWriter` https://docs.rs/parquet/latest/parquet/arrow/async_writer/struct.AsyncArrowWriter.html that requires something to implement `tokio::io::AsyncWrite`.

We could therefore implement `AsyncWrite` on `S3MultiPartUploader` to make this more elegant, though it seems somewhat non-trivial to do so (based on https://itnext.io/asyncwrite-and-a-tale-of-four-implementations-e63aef8397f7 ). I'm tempted to leave that as a TODO / potential future improvement if you agree

Reply:
Yeah just leave it as a todo!